### PR TITLE
Handle HTTP status 428 when sending a message

### LINKED ISSUE #189: Send message HTTP status 428 not handled
Signal may respond to sending a message with a HTTP error 428. It indicates that the user must provide "proof" that they are not a robot. This has affected both new registrations sending the very first messages, and older registrations "out of the blue". A guesstimated list of tasks:

- [x] Catch the HTTP 428
- [x] Pass it to the client
- [x] Receive the completed proof from client
- [x] Send the proof to server
- [x] Tell the client if the proof was accepted

Related issues in Whisperfish:
- [Implement 428 response with a captcha](https://gitlab.com/whisperfish/whisperfish/-/issues/418)
- [Newly registered account gets 428 rate limiter errors](https://gitlab.com/whisperfish/whisperfish/-/issues/417)

### PR DESCRIPTION
**Pull request by direc85**
Fixes #189

### CODE REVIEW

File: libsignal-service/src/sender.rs [hunk 1]
```diff
@@ -538,6 +541,15 @@ where
...
                     }
                 },
+                Err(ServiceError::ProofRequiredError(ref p)) => {
+                    log::debug!("{:?}", p);
+                    if p.options.contains(&String::from("recaptcha")) {
+                        return Err(MessageSenderError::ProofRequired {
+                            token: p.token.clone(),
+                            options: p.options.clone(),
+                        });
+                    }
```
Comment (rubdos) [inline libsignal-service/src/sender.rs:551]: What about the `else` case, then it will just return an `Ok` and silently fail?

**Comment by direc85** [comment]
I believe this is "barely complete enough" to receive proper review. The basic functionality _should_ be there, but because I haven't been hitting the 428 error myself, and submitting the rate limit reCaptcha response requires the token received with the 428 error, this is effectively untested.

File: libsignal-service-actix/src/push_service.rs [hunk 2]
```diff
@@ -107,6 +107,15 @@ impl AwcPushService {
...
             },
+            StatusCode::PRECONDITION_REQUIRED => {
+                let proof_required = response.json().await.map_err(|e| {
+                    log::error!("Failed to decode HTTP 428 response: {}", e);
+                    ServiceError::UnhandledResponseCode {
+                        http_code: StatusCode::PRECONDITION_REQUIRED.as_u16(),
+                    }
+                })?;
+                Err(ServiceError::ProofRequiredError(proof_required))
+            },
```
Comment (rubdos) [inline libsignal-service-actix/src/push_service.rs:118]: We might also want this in the websocket, which requires #191. I don't think this code actually works on Whisperfish, after we started using the WS for sending messages?

Comment (direc85) [reply libsignal-service-actix/src/push_service.rs:118]: I think this is now fixed in [81caabe](https://github.com/whisperfish/libsignal-service-rs/pull/190/commits/81caabe071e11ff3c51c66bb27dbe1fee9a6b5e3)..?

(I still plan to squish the commits for cleaner commit history.)

File: libsignal-service/src/sender.rs (hunk 1 above)
Comment (direc85) [reply libsignal-service/src/sender.rs]: Fixed by not filtering the captcha type here; let the client do it instead.

Comment (rubdos) [reply libsignal-service/src/sender.rs]: Sounds good to me, we can always refine that later.

File: libsignal-service-actix/src/push_service.rs (hunk 2 above)
Comment (direc85) [reply libsignal-service-actix/src/push_service.rs:118]: It's [ca14821](https://github.com/whisperfish/libsignal-service-rs/pull/190/commits/ca14821f02a7371a7bd472bcb436fe4d1a7c0ac1) after cleanup.

**Pull request merged**